### PR TITLE
Start documenting high-level architecture and workflow lifecycle

### PR DESCRIPTION
**Pull request by dandavison**
**What changed?**
- Add a high-level introduction to Temporal Server architecture
- Add @yiminc's Workflow Lifecycle sequence diagrams, with some code entrypoint links. 

**Why?**
- We want to start maintaining documentation of Temporal server internals in GitHub; this PR is intended to start that project.

**Testing/risks**
- N/A: GitHub markdown documentation changes only

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,18 +6,20 @@
...
-Temporal is being developed by [Temporal Technologies](https://temporal.io/), a startup by the creators of Cadence.
+Temporal is a mature technology that originated as a fork of Uber's Cadence.
+It is developed by [Temporal Technologies](https://temporal.io/), a startup by the creators of Cadence.
 
-[![image](https://user-images.githubusercontent.com/12602502/136433917-98abe0d7-4f81-4f97-9b11-62b331c76608.png)](http://www.youtube.com/watch?v=f-18XztyN6c "Temporal")
+[![image](https://user-images.githubusercontent.com/12602502/136433917-98abe0d7-4f81-4f97-9b11-62b331c76608.png)](http://www.youtube.com/watch?v=f-18XztyN6c 'Temporal')
 
 Learn more about Temporal at [docs.temporal.io](https://docs.temporal.io).
 
+To learn about Temporal's internal implementation see [docs/](./docs/README.md).
```
Comment (lorensr) [inline README.md:21]:
```suggestion
Learn more:

- [Courses](https://learn.temporal.io/courses/temporal_101/)
- [Docs](https://docs.temporal.io)
- Internal architecture: [docs/](./docs/README.md)
```

File: README.md [hunk 2]
```diff
@@ -6,18 +6,20 @@
...
+Temporal is a durable execution platform that enables developers to build scalable applications without sacrificing productivity or reliability.
+The Temporal server executes units of application logic called Workflows in a resilient manner that automatically handles intermittent failures, and retries failed operations.
 
-Temporal is a mature technology, a fork of Uber's Cadence.
-Temporal is being developed by [Temporal Technologies](https://temporal.io/), a startup by the creators of Cadence.
+Temporal is a mature technology that originated as a fork of Uber's Cadence.
+It is developed by [Temporal Technologies](https://temporal.io/), a startup by the creators of Cadence.
 
-[![image](https://user-images.githubusercontent.com/12602502/136433917-98abe0d7-4f81-4f97-9b11-62b331c76608.png)](http://www.youtube.com/watch?v=f-18XztyN6c "Temporal")
+[![image](https://user-images.githubusercontent.com/12602502/136433917-98abe0d7-4f81-4f97-9b11-62b331c76608.png)](http://www.youtube.com/watch?v=f-18XztyN6c 'Temporal')
```
Comment (lorensr) [inline README.md:17]:
```suggestion
[![image](https://github.com/temporalio/temporal/assets/251288/693d18b5-01de-4a3b-b47b-96347b84f610)](https://youtu.be/wIpz4ioK0gI 'Getting to know Temporal')
```

**Review by lorensr — COMMENTED** [review]
Love it! 🙏

File: README.md (hunk 1 above)
Comment (dandavison) [reply README.md]: Thanks, I like that. Committed.

File: README.md (hunk 2 above)
Comment (dandavison) [reply README.md]: Committed.

**Review by MichaelSnowden — COMMENTED** [review]
This looks nice. One nit I have is that I don't see us referring to task queues as matching shards anywhere else. Conceptually, it's similar in some ways to sharding. However, I think that it might confuse some operators because they need to set the number of shards, and they might ask misguided questions like, "how do I control the number of matching shards?"

**Pull request closed without merging**